File: examples/quickstart.ts
/**
 * orderflow-metrics — quickstart (TypeScript / Node)
 *
 * A self-contained tour: deterministic synthetic market data in, a spread of
 * microstructure metrics out. No external data, no dependencies.
 *
 *   node --experimental-strip-types examples/quickstart.ts
 */
import {
  ofi,
  tradeImbalance,
  bucketByVolume,
  vpin,
  corwinSchultz,
  abdiRanaldo,
  parkinsonVolatility,
  yangZhangVolatility,
  realizedVolatility,
  realizedSkewness,
  realizedKurtosis,
  varianceRatio,
  hurstExponent,
  type L1Quote,
  type Trade,
  type Candle,
} from "orderflow-metrics";

// Deterministic pseudo-random stream (MINSTD) so the numbers reproduce exactly.
let seed = 1;
const rand = () => (seed = (48271 * seed) % 2147483647) / 2147483647;

// 1) Best-quote updates -> Order Flow Imbalance
const quotes: L1Quote[] = [
  { bidPrice: 100, bidSize: 5, askPrice: 101, askSize: 4 },
  { bidPrice: 100, bidSize: 8, askPrice: 101, askSize: 2 },
  { bidPrice: 100.5, bidSize: 3, askPrice: 101, askSize: 1 },
];
console.log("OFI:                  ", ofi(quotes));

// 2) Trades -> trade imbalance + VPIN (flow toxicity)
const trades: Trade[] = Array.from({ length: 400 }, () => {
  const r = rand();
  return {
    price: 100 + (r - 0.5),
    size: 1 + Math.floor(rand() * 5),
    side: r > 0.5 ? "buy" : "sell",
  } as Trade;
});
console.log("Trade imbalance:      ", tradeImbalance(trades).toFixed(4));
console.log("VPIN (flow toxicity): ", vpin(bucketByVolume(trades, 200), { window: 10 }).toFixed(4));

// 3) Daily OHLC candles -> spread estimators + range-based volatility
const candles: Candle[] = [];
let px = 100;
for (let i = 0; i < 60; i++) {
  const open = px;
  const close = open + (rand() - 0.5) * 1.5;
  const high = Math.max(open, close) + rand() * 0.8;
  const low = Math.min(open, close) - rand() * 0.8;
  candles.push({ open, high, low, close });
  px = close;
}
console.log("Corwin-Schultz spread:", corwinSchultz(candles).toFixed(5));
console.log("Abdi-Ranaldo spread:  ", abdiRanaldo(candles).toFixed(5));
console.log("Parkinson vol:        ", parkinsonVolatility(candles).toFixed(5));
console.log("Yang-Zhang vol:       ", yangZhangVolatility(candles).toFixed(5));

// 4) Return series -> realized measures, higher moments, efficiency, long memory
const returns = candles.slice(1).map((c, i) => Math.log(c.close / candles[i].close));
console.log("Realized vol:         ", realizedVolatility(returns).toFixed(5));
console.log("Realized skewness:    ", realizedSkewness(returns).toFixed(4));
console.log("Realized kurtosis:    ", realizedKurtosis(returns).toFixed(4));
console.log("Variance ratio (q=4): ", varianceRatio(returns, 4).toFixed(4));
console.log("Hurst exponent:       ", hurstExponent(returns).toFixed(4));

File: examples/quickstart.py
"""orderflow-metrics — quickstart (Python)

A self-contained tour: deterministic synthetic market data in, a spread of
microstructure metrics out. No external data, no dependencies.

    python examples/quickstart.py
"""
import math

from orderflow_metrics import (
    Candle,
    L1Quote,
    Trade,
    abdi_ranaldo,
    bucket_by_volume,
    corwin_schultz,
    hurst_exponent,
    ofi,
    parkinson_volatility,
    realized_kurtosis,
    realized_skewness,
    realized_volatility,
    trade_imbalance,
    variance_ratio,
    vpin,
    yang_zhang_volatility,
)

# Deterministic pseudo-random stream (MINSTD) so the numbers reproduce exactly.
_seed = 1


def rand() -> float:
    global _seed
    _seed = (48271 * _seed) % 2147483647
    return _seed / 2147483647


# 1) Best-quote updates -> Order Flow Imbalance
quotes = [
    L1Quote(bid_price=100, bid_size=5, ask_price=101, ask_size=4),
    L1Quote(bid_price=100, bid_size=8, ask_price=101, ask_size=2),
    L1Quote(bid_price=100.5, bid_size=3, ask_price=101, ask_size=1),
]
print("OFI:                  ", ofi(quotes))

# 2) Trades -> trade imbalance + VPIN (flow toxicity)
trades = []
for _ in range(400):
    r = rand()
    trades.append(
        Trade(price=100 + (r - 0.5), size=1 + int(rand() * 5), side="buy" if r > 0.5 else "sell")
    )
print("Trade imbalance:      ", round(trade_imbalance(trades), 4))
print("VPIN (flow toxicity): ", round(vpin(bucket_by_volume(trades, 200), window=10), 4))

# 3) Daily OHLC candles -> spread estimators + range-based volatility
candles = []
px = 100.0
for _ in range(60):
    open_ = px
    close = open_ + (rand() - 0.5) * 1.5
    high = max(open_, close) + rand() * 0.8
    low = min(open_, close) - rand() * 0.8
    candles.append(Candle(open_, high, low, close))
    px = close
print("Corwin-Schultz spread:", round(corwin_schultz(candles), 5))
print("Abdi-Ranaldo spread:  ", round(abdi_ranaldo(candles), 5))
print("Parkinson vol:        ", round(parkinson_volatility(candles), 5))
print("Yang-Zhang vol:       ", round(yang_zhang_volatility(candles), 5))

# 4) Return series -> realized measures, higher moments, efficiency, long memory
returns = [math.log(candles[i].close / candles[i - 1].close) for i in range(1, len(candles))]
print("Realized vol:         ", round(realized_volatility(returns), 5))
print("Realized skewness:    ", round(realized_skewness(returns), 4))
print("Realized kurtosis:    ", round(realized_kurtosis(returns), 4))
print("Variance ratio (q=4): ", round(variance_ratio(returns, 4), 4))
print("Hurst exponent:       ", round(hurst_exponent(returns), 4))
